lib/srv/desktop/rdp/rdpclient/src/rdpdr.rs

```diff
@@ -4173,32 +4178,101 @@ mod tests {
         tpkt::Payload::Raw(p)
     }
 
-    #[test]
-    fn test_server_announce() {
-        let mut c = client();
-        // Incoming payload of:
-        // SharedHeader { component: RDPDR_CTYP_CORE, packet_id: PAKID_CORE_SERVER_ANNOUNCE }
-        // ServerAnnounceRequest { version_major: 1, version_minor: 13, client_id: 3 }
-        let payload = create_payload(
+    fn test_payload(c: &mut Client, payload_in: Vec<u8>, payload_out: Messages) {
+        let payload_in = create_payload(payload_in, 10);
+        assert_eq!(c.read_and_create_reply(payload_in).unwrap(), payload_out)
+    }
+
+    /// Incoming payload of:
+    /// SharedHeader { component: RDPDR_CTYP_CORE, packet_id: PAKID_CORE_SERVER_ANNOUNCE }
+    /// ServerAnnounceRequest { version_major: 1, version_minor: 13, client_id: 3 }
+    ///
+    /// Response payload of:
+    /// ClientAnnounceReply ClientIdMessage { version_major: 1, version_minor: 12, client_id: 3 }
+    /// ClientNameRequest { unicode_flag: Ascii, computer_name: "teleport" }
+    fn test_handle_server_announce(c: &mut Client) {
+        test_payload(
+            c,
             vec![
                 2, 240, 128, 104, 0, 1, 3, 236, 240, 20, 12, 0, 0, 0, 3, 0, 0, 0, 114, 68, 110, 73,
                 1, 0, 13, 0, 3, 0, 0, 0,
             ],
-            10,
+            vec![
+                vec![
+                    12, 0, 0, 0, 3, 0, 0, 0, 114, 68, 67, 67, 1, 0, 12, 0, 3, 0, 0, 0,
+                ],
+                vec![
+                    25, 0, 0, 0, 3, 0, 0, 0, 114, 68, 78, 67, 0, 0, 0, 0, 0, 0, 0, 0, 9, 0, 0, 0,
+                    116, 101, 108, 101, 112, 111, 114, 116, 0,
+                ],
+            ],
+        );
```

Comment:
Can you think of a way to write this so that the logic that is being tested is more apparent? These arbitrary vecs are impossible to reason about (which is why you have the comments above the test).

Do these structs have an encode as well? Maybe if we encode a PDU object instead of hard coding its bytes it will be easier to see what's going on?

For reference, the cliprdr tests make it a little more obvious what's supposed to be happening. Here's an example:

```
    fn responds_to_format_data_request_hasdata() {
        // a null-terminated utf-16 string, represented as a Vec<u8>
        let test_data = util::to_unicode("test", true);

        let mut c: Client = Default::default();
        c.clipboard
            .insert(ClipboardFormat::CF_UNICODETEXT as u32, test_data.clone());

        let req = FormatDataRequestPDU::for_id(ClipboardFormat::CF_UNICODETEXT as u32);
        let responses = c
            .handle_format_data_request(&mut Cursor::new(req.encode().unwrap()))
            .unwrap();

        // expect one FormatDataResponsePDU
        assert_eq!(responses.len(), 1);
        let mut payload = Cursor::new(responses[0].clone());
        let _pdu_header = vchan::ChannelPDUHeader::decode(&mut payload).unwrap();
        let header = ClipboardPDUHeader::decode(&mut payload).unwrap();
        assert_eq!(header.msg_type, ClipboardPDUType::CB_FORMAT_DATA_RESPONSE);
        assert_eq!(header.msg_flags, ClipboardHeaderFlags::CB_RESPONSE_OK);
        assert_eq!(header.data_len, 10);
        let resp = FormatDataResponsePDU::decode(&mut payload, header.data_len).unwrap();
        assert_eq!(resp.data, test_data);
    }
```

Instead of arbitrary byte sequences here, you can see that we're calling `handle_format_data_request` with a request, decoding the response, and verifying that it looks correct.

Reply:
The ideal that I'm aiming for with this test suite is: pass this `rdpdr::Client` data precisely how we receive it from the RDP server, and check that it gets processed into precisely the data it should send back to the RDP server, for many of the common scenarios we encounter while manually using the feature. In concrete terms this means testing that the `message` we pass in here is converted to the `responses` we send back.

https://github.com/gravitational/teleport/blob/8b488c9861008a68a8232389bd5245b94c8eadf4/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L635

If I achieve that, it means that I can do whatever sort of refactor I want to within `rdpdr::Client`, and be confident that it hasn't broken anything important by simply running the test suite. We also get test coverage for the `rdpdr::Client`'s relatively complex internal state, and for all of the many individual `encode` and `decode` functions, without needing to write specific unit tests for each of them.

I did have a previous iteration of `test_handle_server_announce` that looked like [this](https://github.com/gravitational/teleport/pull/16167/commits/90953f9ae4cd7e09272f13ab42560d4bcaaa7e61#diff-649de9ed3b460c7d074083e126f424535b1c48bcd9b6a881fa278c66da107c94R4223-R4281
), and where I'd added a new trait [`Encode`](https://github.com/gravitational/teleport/pull/16167/commits/90953f9ae4cd7e09272f13ab42560d4bcaaa7e61#diff-04d4ba71603f3d244046f417a7b96c4830840c73c3d18a361b83364095ee2f3cR1872-R1875) along with the [`downcast-rs`](https://crates.io/crates/downcast-rs) library, which made the tests more comprehensible than using these raw vectors. But I moved away from that because I realized that I was rebuilding some of the internal logic to `rdpdr::Client` in the test suite, and discovered that that was unnecessary with the approach described above.

I think I see a way in which we can get the best of both worlds, I'll make this PR a draft and branch off to take a stab at it (unless you cut me off for being too idealistic 🙂)

Reply:
https://github.com/gravitational/teleport/pull/16299